foundation/org.eclipse.persistence.core/resource/org/eclipse/persistence/internal/helper/VendorNameToPlatformMapping.properties

```diff
@@ -32,6 +34,10 @@
 (?is)oracle.*11.*=org.eclipse.persistence.platform.database.oracle.Oracle11Platform
 (?is)oracle.*10.*=org.eclipse.persistence.platform.database.oracle.Oracle10Platform
 (?is)oracle.*9.*=org.eclipse.persistence.platform.database.oracle.Oracle9Platform
+(?is)core.oracle.*12.*=org.eclipse.persistence.platform.database.Oracle12Platform
+(?is)core.oracle.*11.*=org.eclipse.persistence.platform.database.Oracle11Platform
```

Comment:
Perhaps I don't understand what you mean by these lines being duplicate and unneeded.
These reg expressions are used by the platform detection code (See ```org.eclipse.persistence.internal.helper.DBPlatformHelper.getDBPlatform```).
If there are no extension, OracleXPlatform classes available to classload then the default is just OraclePlatform.class.

Reply:
But I'm not proposing to move platform files from '.oracle' to '.core'. This PR is only ADDING the .core OracleXPlatform classes so they can be detected in the even the extension bundle is not available. There is no functional change for the normal usecase of the extension bundle being available. In that case, the .oracle platform classes will be detected and used as normal. 
Granted, this PR is proposing moving the NString, NClob, and NCharacter classes into core, but I was only doing that cause it looked to me like they were unnecessarily in the extension project. I can remove that part of the PR if it matters.

Reply:
@lukasj Thoughts?

Reply:
I'm off (not working and not on vacation :-/) this week (since last Thu) - let me get back to this next week. Thx!

Reply:
I missed it was just a copy, not move, my bad :-/ In any case I don't like having duplicates - is there some way we could avoid that?

Reply:
@lukasj I agree, duplication is not normally preferable and I am open to suggestions, but I find this implementation cleaner than alternatives I attempted (like adding a version field to OraclePlatform). 
The issue is that the OracleXPlatform classes don't completely depend on the Oracle extension libs. There is impl in there that could/should be available in core. Since those XPlatform classes are locked out in the extension bundle, the default of OraclePlatform (essentially Oracle 8 support from 1998) is just too old. Perhaps an alternative is to start eliminating those older platform classes and move up the default implementation (OraclePlatform) in core.

Reply:
@lukasj The normal use case involves having **org.eclipse.persistence.oracle** available and for this case, those platform classes will still take precedent. This only adds support for environments that are running without the **org.eclipse.persistence.oracle** bundle. Currently, in such an environment, we throw away the version information in favor of using just one base OraclePlatform class. At the very least, I feel there is value in having OracleXPlatform classes available in core to accurately portray the Oracle version.

Reply:
I would really like to make some progress on this issue. While I agree with you that duplication is not normally a good design decision, I feel like the **org.eclipse.persistence.oracle** bundle and it's runtime dependencies has forced the requirement for us to have a more reasonable fallback option available in core. The fallback option of reverting to Oracle v8 support is not good enough to me. 
I think simply having the empty OracleXPlatform classes alone would be a valuable improvement to the core bundle since it preserves the version information. From that step forward, I see no issues with porting some support to these new platform classes. Different environments would be running with one set over another, so I think duplication is minimum.

Reply:
For master:

I took a look into Oracle DB support policy, which is available at http://www.oracle.com/us/support/library/lifetime-support-technology-069183.pdf and according to the info I've found there, I would personally be more drastic when in comes to what to do for master which should become 3.0 given that major version change allow us to "break" backwards compatibility ;-)

* v8 version support ended in Dec 2006 => drop
* v9 version support ended in Jul 2010 => drop
* v10 version support ended in Jul 2013 => drop
* v11 version support ends in Dec 2020 => keep it (probably easier to do then checking DB version for v11.1 for which support ended in Aug 2015)
* v12 support ends in Mar 2025

This should allow us to move some code to the base OraclePlatform class and get rid of some  already deprecated `oracle.sql` usages. Probably also define some `OracleProvider` interface allowing us to hook oracledb specific stuff into one common OraclePlatform impl (just thinking loud).

For 2.7 I don't think we can be that "drastic", so even though I don't like this, I could live with it as I do not have any alternative :-/

Reply:
@lukasj 
I was under the impression that Oracle had a "lifetime" support statement for the Oracle versions and that the likely-hood of us sunsetting v8-v10 on any EclipseLink version was extremely low! I am in support of condensing down some of this support on master (eventually that will be shipped in Liberty).
Duplication is not normally a good design decision since it only leads to more overhead when it comes to code changes; making code maintainability harder. For this change in particular, I too do not have any better of an alternative and even so, I feel it's fairly small/low impact. I think the majority of users (especially in SE environments) will simply have the **org.eclipse.persistence.oracle** bundle available and these core platforms will not be touched. The main benefit here is that it provides a mechanism for us to port support to Oracle DB users that don't/can't have the **org.eclipse.persistence.oracle** bundle available.

On a side note for the review: 

 * I removed the change that pulled 'NCharacter', 'NClob', and 'NString' into core. I had figured since those classes don't have any hard deps on the JDBC driver, maybe I could just move them over. I then felt it might be muddying the real change I was advocating so I pulled it out.
 * Updated some comments

If you agree with all the changes or think I should change something, let me know. I would like to backport this to 2.7 as well.

Reply:
Even "lifetime" support apparently has its limitations :-) based on the policy I shared those who will need support for v8-v10 should go through sustaining (or find a way to reuse the source from previous eclipselink versions - up to 2.7.x) or just ask us and we may deal with it but we should not carry the graveyard with us forever. To be clear - this is my personal point of view.

I believe we want to free our hands at some point, simplify the code and adopt newer APIs and 3.0 is good opportunity to do so. If we can make it easier for Liberty to include better Oracle DB specific support out of the box, even better and let's do it.

Reply:
Speaking of 3.0... are we still planning on cutting that version at the schedule laid out here: https://projects.eclipse.org/projects/ee4j.eclipselink ? If so, I don't think we will be able to get this change in by tomorrow ;)

Reply:
nope, I'll have to update that page. I believe it would be good to have 3.0 properly running on JDK 9, so its date may be moved after the release date of JDK 11 as we will very likely need to do some backwards incompatible changes for that (like dealing with removal of corba from JDK - I assume we'll get PR for this soon..)